src/sessions/PageManager.ts

```diff
@@ -156,10 +339,68 @@ export class PageManager {
         this.record(PAGE_VIEW_TYPE, this.createPageViewEvent());
     }
 
+    private moveItemsFromBuffer = (item: any) => {
+        this.page.ongoingActivity.add(item);
+    };
+
     /**
      * Returns true when cookies should be used to store user ID and session ID.
      */
     private useCookies() {
         return navigator.cookieEnabled && this.config.allowCookies;
     }
+
+    /**
+     * Checks whether the virtual page is still being loaded.
+     * If completed:
+     * (1) Clear the timers
+     * (2) Record data using NavigationEvent
+     * (3) Indicate current page has finished loading
+     */
+    private periodicCheckerCallback = () => {
+        if (this.page.ongoingActivity.size === 0 && this.fetchCounter === 0) {
+            clearInterval(this.periodicCheckerId);
+            clearTimeout(this.activityTimeoutCheckerId);
+            this.domMutationObserver.disconnect();
+            this.recordVirtualPageNavigationEvent();
+            this.periodicCheckerId = undefined;
+            this.activityTimeoutCheckerId = undefined;
+            this.page.isLoaded = true;
+        }
+    };
+
+    /** Clears timers and disconnects observer to stop page timing. */
+    private activityTimeoutCallback = () => {
+        clearInterval(this.periodicCheckerId);
+        this.periodicCheckerId = undefined;
+        this.activityTimeoutCheckerId = undefined;
+        this.domMutationObserver.disconnect();
+        this.page.isLoaded = true;
+    };
+
+    /** Resets periodic check timer and updates page's latestEndTime. */
+    private domMutationCallback = () => {
+        this.page.latestEndTime = Math.max(this.page.latestEndTime, Date.now());
```

Comment:
question: Under what conditions would `this.page.latestEndTime` be greater than `Date.now()`?